Step5 - 로또(수동) PR 요청드립니다.

haservi opened the pull request:
안녕하세요. 선흠님

Step4 말씀해주신 코멘트에 따라 코드 수정 및 Step5 로또(수동) 기능 추가했습니다.

감사합니다.

haservi committed:
refactor(Step4): 코멘트 요청 사항 적용 및 코드 정리

- 로또 티켓과 당첨 번호 비교한 값을 변수로 저장
- 테스트 코드 정리
- WinningNumbers 사용하지 않은 생성자 제거

haservi committed:
chore(LottoRank): 변수이름과 함수이름 변경

haservi committed:
docs(README): 로또(수동) 요구사항 정리

haservi committed:
test(Input): 구매 금액 입력 시 유효성 검사 테스트 코드 추가

haservi committed:
refactor(LottoGenerator): 로또 생성기 interface로 변경

haservi committed:
refactor(CustomLotto): 수동 로또 입력위해 구조 변경 및 클래스 역할 변경 및 테스트 코드 추가

haservi committed:
feat(CustomLotto): 수동 로또 기능 추가 및 코드 정리

haservi committed:
feat(ResultView): 문구 수정 및 코드 정리

haservi committed:
feat(CustomLotto): 수동 로또 예외 처리 및 코드 정리

shared-moon commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -15,19 +16,37 @@ public LottoController(LottoGame lottoGame) {
 
-        ResultView.lottoPurchase(lottoTickets.ticketCount(), lottoTickets.toString());
+        LottoMoney lottoMoney = InputView.getLottoPurchasePrice();
+        int customLottoTicketCount = InputView.getCustomLottoCount(lottoMoney);
```
큰 상관은 없지만 보편적으로 수동을 표현할 땐 Manual이라는 표현을 많이 써요!

shared-moon commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -9,37 +9,90 @@ public class InputView {
+        System.out.println(ENTER_CUSTOM_LOTTO_NUMBER);
+    }
+
+    public static LottoGenerator getCustomLottoNumbers(int count, int total) {
```
여기 약간 결합도가 높아지는 부분인 것 같은데요, view에서는 응답을 가져다 쓰는 주체가 무엇인지 알 필요가 없지 않을까요 ?

haservi replied:
네 여기서 결합도가 증가한 부분이 저도 고민이 된 부분이에요.

프로그램 요구사항에서 **예외 처리를 통해 에러가 발생하지 않도록 한다**. 라는 문구가 있어서..
기존 로직 try catch처리 하기 위해 view에 썻네요,,

https://cheese10yun.github.io/spring-guide-exception/ 이전 스프링 프로젝트 중에는 이분 블로그를 참고하여 error response를 처리 했었습니다. 

중간에 dto를 둬도 좋은 방법이 잘 안떠올라 저렇게 로직 구현을 한 부분입니다. ㅠ
혹시 깔끔하게 처리하는 방법이 있을까요??

shared-moon commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -13,26 +11,24 @@ public LottoGame() {
-        LottoGenerator lottoGenerator = new LottoGenerator();
         List<LottoTicket> lottoTicketList = new ArrayList<>();
-        for (int i = 0; i < ticketCount; i++) {
+        for (LottoGenerator lottoGenerator : lottoGeneratorList) {
```
아쉽게도 사용처가 추상화를 해서 얻는 이점이 크게 없어보이네요..!

단편적인 예가 될 것 같긴 한데, 지금 상황에서 보면 LottoGenerator 인터페이스가 위아래로 사용되고 있어 각각이 뭐가 다른지에 대해 한 단계 더 고민하게 만드는 모양새인 것 같아요!

추상화가 나쁘다는 건 아니지만, 추상화를 사용할때는 분명히 트레이드오프가 있는데요, 이번만큼은 약간 손해가 좀 더 있지 않나 하는 생각입니다! 어떻게 생각하시나요 ? :)

shared-moon replied:
꼭 고치라는 의미는 아닙니다! 그냥 한번 생각 해 보는 시간을 가져보자는 의미였어요 :)

haservi replied:
로또 게임이 수동과 자동을 함께 입력 가능하여 수동 입력 후 해당값을 매개변수로 넘겨서 한꺼번에 생성하려는 용도였는데 
한번 다른 방법을 고민해보겠습니다.

shared-moon commented on src/main/java/lotto/domain/LottoMoney.java:
```diff
@@ -0,0 +1,84 @@
+        return ticketCount;
+    }
+
+    public int minusTicketCount(int count) {
```
메서드명이 뭔가 감산만 할것 같은 느낌이네요!

shared-moon replied:
약간 `AtomicInteger` 클래스를 보시면 제가 얘기하고싶은 메서드 네이밍의 효과가 뭔지 잘 보일 거 같아서 남겨봅니다!

haservi replied:
감사합니다.

shared-moon reviewed (changes requested):
하섭님 안녕하세요!
5단계 미션 잘 진행 해 주셨네요 :)
코멘트를 몇 개 남겨 두었는데요, 그 중에서 패키지 의존성 관련해서는 한 번 정리 해 보고 넘어가는 게 좋을 것 같다는 생각이 들어서 한 번 확인 해 보시고 리뷰 재 요청 부탁드릴게요!
혹시 궁금한 점 있으시면 언제든지 DM 주세요!
화이팅입니다 🔥

haservi committed:
style(Variable): 수동 로또 이름 custom -> manual로 변경

haservi committed:
style(Method): 함수 이름 변경 및 출력 형식 변경

haservi committed:
refactor(Dto): input 영역 반환 타입을 Dto로 변경

shared-moon commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -9,37 +12,89 @@ public class InputView {
+        System.out.println(ENTER_MANUAL_LOTTO_NUMBER);
+    }
+
+    public static LottoManualGeneratorRequestDto getManualLottoNumbers(int count, int total) {
```
로그에도 필수적인 정보는 아닌 것 같은데 total을 꼭 넘겨야 할까요 ?

haservi replied:
수동 입력 실패시 사용자가 얼마나 입력이 남았는지 보여줄 용도로 콘솔창에 보여주려 했는데 
좋은 방법이 아닌걸까요?? '';

shared-moon replied:
의도는 좋다고 생각합니다!
하지만 total을 넘김으로써 메서드 시그니처가 조금 어지러워지는 경향이 있어 꼭 필요한 부분이 아니라면 제외하는게 어떨까 싶었어요!

shared-moon commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -15,19 +19,38 @@ public LottoController(LottoGame lottoGame) {
+    private void buyLotto(LottoMoney lottoMoney, int manualTicketCount, List<LottoGenerator> lottoGeneratorList) {
+        LottoTickets lottoTickets = lottoGame.buy(lottoMoney, lottoGeneratorList);
+        ResultView.lottoPurchase(manualTicketCount, lottoTickets.autoTicketCount(manualTicketCount), lottoTickets.toString());
+        WinningLottoNumbers winningNumbers = InputView.getWinningNumbers().getWinningLottoNumbers();
```
buyLotto에서 winningLottoNumbers를 받는 역할을 하는게 맞을까요 ?

haservi replied:
그렇네요. 해당 함수 역할과 관련이 없는 부분이라 해당 함수에서 빼는게 좋을 것 같아요.

shared-moon commented on src/main/java/lotto/domain/AutoLottoGenerator.java:
```diff
@@ -0,0 +1,34 @@
+        return lottoNumbers;
+    }
+
+    private static List<Integer> availableLottoRange() {
```
매번 45개짜리 리스트를 새로 생성해야 할 것 같은데요, 미리 만들어두고 재사용하는 건 어떨까요 ?

haservi replied:
네 그게 좋은 것 같아요..!

shared-moon commented on src/main/java/lotto/domain/LottoTickets.java:
```diff
@@ -14,17 +14,22 @@ public LottoTickets(List<LottoTicket> lottoTicketList) {
 
     public int ticketListPrice() {
-        return this.lottoTicketList.size() * LOTTO_PRICE;
+        return lottoTicketList.size() * LOTTO_PRICE;
```
현재 테스트에서만 쓰이는 것 같은데, 지워도 괜찮지 않을까요 ?

haservi replied:
넵

shared-moon commented on src/main/java/lotto/dto/LottoManualGeneratorRequestDto.java:
```diff
@@ -0,0 +1,16 @@
+
+import lotto.domain.ManualLottoGenerator;
+
+public class LottoManualGeneratorRequestDto {
```
약간 DTO로 전달할 데이터로는 적합하지 않아보이는데, 혹시 DTO 개념을 적용하신 이유가 있을까요 ?

haservi replied:
이전에 코멘트 주신 결합도가 높아 보인다 하셔서 dto로 구현은 했는데..
전 처음에 작성했던 반환값이 사용자 요구사항에 맞춘 비즈니스 로직이라는 생각입니다 '';

반환값을 결합도를 낮춰 String이나 int로 받으면 프로그램이 죽지 않고 예외처리하는 방법을 잘 모르겠습니다.
이 부분은 코멘트 한번 부탁드려요. 🙏

shared-moon commented on src/main/java/lotto/enums/LottoRank.java:
```diff
@@ -21,19 +21,19 @@ public enum LottoRank {
 
     private static boolean matchLottoRank(LottoRank lottoRank, int matchCount, boolean bonus) {
-        return lottoRank.matchCount == matchCount && lottoRank.bonus == isBonus(matchCount, bonus);
+        return (lottoRank.matchCount == matchCount) && (lottoRank.bonus == isBonus(matchCount, bonus));
```
이 메서드 혹시 굳이 static이 아니어도 되지 않을까요 ?

haservi replied:
음 위쪽 findLottoRank가 static으로 해당 함수를 사용해서 static이 아니면 안되는걸로 이해하는데
좋은 방법이 있을까요??

shared-moon replied:
일반 메서드로 제공하고 위 기준으로 `item.matchLottoRank(matchCount, bonus)` 처럼 시그니처를 바꿔도 될 것 같아서요!

shared-moon commented on src/test/java/lotto/domain/LottoMoneyTest.java:
```diff
@@ -0,0 +1,38 @@
+
+    @DisplayName("로또 구매 실패 확인")
+    @ParameterizedTest(name = "\"{0}\"원은 로또 구매를 할 수 없습니다.")
+    @ValueSource(strings = {"", " ", "100", "-1", "10004", "1000a"})
```
이건 취향이긴 한데, 저는 `@MethodSource`를 주로 사용합니다 !

shared-moon reviewed (changes requested):
하섭님 안녕하세요!
5단계 미션 잘 진행 해 주셨네요 :)
코멘트를 몇 개 남겨 두었는데, 한 번만 더 같이 보면 좋을 거 같아서 확인하시고 리뷰 재요청 부탁드릴게요!
화이팅입니다 🔥

haservi committed:
refactor(Comment): 코멘트 내용 반영하여 코드 수정

- 로또 생성 범위 함수는 한번만 사용하도록 변경
- buyLotto함수와 관련이 없는 코드 분리
- 테스트에서만 사용하는 코드 제거 및 테스트 코드 추가
- dto 제거(해당 코드 원복 및 다른 방법으로 구현)

haservi committed:
refactor(InputView): input쪽에서 비즈니스 로직 반환값 제외 및 유효성 처리

haservi committed:
style(InputView): 수동 입력 실패 시 잔여 횟수 콘솔 제거

haservi committed:
refactor(InputView): Input view 반환값은 String 또는 int만 받도록 변경

shared-moon commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -14,41 +9,35 @@ public class InputView {
-            System.out.printf((ERROR_PREFIX) + "%n", e.getMessage());
-            return manualPurchaseCount(lottoMoney);
-        }
+    public static String getManualLottoNumbers() {
```
👍

shared-moon reviewed (approved):
하섭님 안녕하세요!
피드백 반영 잘 해 주셨네요 :)
코멘트를 몇 개 남겨 두었는데, 다음 미션 진행 전 확인 부탁드립니다!
의존성 관련해서 약간 제 의견을 강요한 것 같은 느낌이 있는데요, 다음 미션 진행 전 한번 생각 해 보시면 좋을 것 같아서 제안 드려 봤어요!
제 의견이 정답이라는 건 아니고, 그저 한번 더 생각 해 보는 시간이 되었길 바랍니다!
이번 미션은 여기서 머지 하도록 하겠습니다 :)
남은 미션도 꼭 끝까지 완수하시길 바래요 🔥 
추가로 문의사항이 있으시면 언제든지 DM 주세요!
